Add Approval revoke + OPERATOR to APPROVED + set_approval_for_all fixes

gRoussac opened the pull request:
This PR is for change of approve and approve_for_all entry point.

set_approve / revoke https://github.com/casper-ecosystem/cep-78-enhanced-nft/issues/155 == https://github.com/casper-ecosystem/cep-78-enhanced-nft/issues/174

set_approval / revoke_approve_all main fix #172 

https://github.com/casper-ecosystem/cep-78-enhanced-nft/issues/176 testing

code duplication #159 

multiple operators #175

davidatwhiletrue committed:
Update CES to v0.2.0
Change `token_id` field type in the events from `Any` to `String.

gRoussac committed:
replaced hardcoded strings with constants

# Conflicts:
#	tests/src/mint.rs

gRoussac committed:
As discussed with @gRoussac I'm changing test support  method name from `call_entry_point_with_ret` to `call_session_code_with_ret` since it does not call directly a contract by `entry_point` and instead it runs session code to call the contract indirectly.

# Conflicts:
#	tests/src/mint.rs

davidatwhiletrue committed:
fix formatting issues

gRoussac committed:
Set TOKEN_ISSUERS entry to token minter instead of token owner (#170)

* Set TOKEN_ISSUERS entry to token minter instead of token owner
Add test on TOKEN_ISSUERS with minter <> owner

* Missing test

gRoussac committed:
Merge branch 'feat-cesp' of github.com:casper-ecosystem/cep-78-enhanced-nft into backport

gRoussac committed:
Rename OPERATOR to APPROVED

gRoussac committed:
Dedup constants test/contract

gRoussac committed:
Update consts

gRoussac committed:
Update readme

gRoussac commented:
@ACStoneCL I will come back to you later for your help to document those changes on approve and approve_all entry points

For now
- introducing a revoke entry point
- renaming operator dict to approved dict (or spender) for approvals storage
- renaming operator argument to spender for approve the entry point
...
later 
- introducing an operators dict for operators storage
- adding a getter entry point :  is approved for all

ACStoneCL commented:
> @ACStoneCL I will come back to you later for your help to document those changes on approve and approve_all entry points
> 
> For now
> 
> * introducing a revoke entry point
> * renaming operator dict to approved dict for approvals storage
> * renaming operator argument to spender for the approve entry point
>   ...
>   later
> * introducing an operators dict for operators storage
> * adding a getter entry point :  is approved for all

@gRoussac Sure thing, I'll take a look and see what I can document as is. If anything is confusing, I'll reach out to make sure it's clear before I put it in writing.

gRoussac committed:
Introducing ARG_SPENDER

gRoussac committed:
Add operators and remove deprec. arg

gRoussac committed:
small change

gRoussac committed:
set_approval_for_all without tests

gRoussac committed:
empty readme client isApprovedForAll

gRoussac committed:
// naming entry point

gRoussac committed:
naming + revoke as operator

gRoussac committed:
// comments out

gRoussac committed:
Add some test for approvals

gRoussac committed:
// small change

gRoussac committed:
typo

gRoussac committed:
// TODO

gRoussac committed:
// naming

gRoussac committed:
Adding create_funded_dummy_account in tests support

gRoussac committed:
Fulfill should_set_approval_for_all test

gRoussac committed:
Readme for new is_approved_for_all client

gRoussac committed:
Add tests for self approvals

gRoussac committed:
Merge branch 'feat-cesp' into approve_revoke

# Conflicts:
#	contract/src/error.rs
#	contract/src/main.rs
#	tests/src/upgrade.rs

gRoussac committed:
test for set_approval_for_all CEP47 event

gRoussac commented:
@davidatwhiletrue @ihor Thanks again for your help for the specs, you can start reviewing now (and eventually run E2E tests with this wasm maybe if you have that in stock).

gRoussac committed:
Set serde-json-wasm as dep (#179)

* Set serde-json-wasm as dep

* backport option import feat_cesp
# Conflicts:
#	client/updated_receipts/src/main.rs
#	contract/Cargo.toml
#	contract/src/main.rs

ACStone-MTS committed:
Readme Update (#4)

gRoussac committed:
Merge cargo.tmol changes

gRoussac commented:
Once reviewed this PR will be squashed into branch feat-cesp.

gRoussac committed:
Update some crates in tests too

gRoussac committed:
// small change

gRoussac committed:
Simplify std for tests

gRoussac committed:
small change

gRoussac committed:
// wording

gRoussac committed:
Merge branch 'approve_revoke' of github.com:gRoussac/cep-78-enhanced-nft into approve_revoke

davidatwhiletrue commented on contract/src/main.rs:
```diff
@@ -807,54 +813,61 @@ pub extern "C" fn approve() {
     ) {
-        Some(token_owner) => token_owner,
+        Some(owner) => owner,
         None => runtime::revert(NFTCoreError::InvalidAccountHash),
```
Should this error be `InvalidTokenIdentifier`?

gRoussac replied:
Hi, so this error says owner not found for this particular token identifier. It is already [this error](https://github.com/casper-ecosystem/cep-78-enhanced-nft/blob/feat-cesp/contract/src/main.rs#L822) in the feat-cesp branch, so I am keeping it like that for both occurrences approve/revoke. To me this is correct behavior to say the owner is not found, thus his account hash is not found as a result. May you find this incorrect then let's raise another issue on dev cesp-branch for that.

davidatwhiletrue replied:
Ok, makes sense to raise this in a separate issue. My concern was that the same error could be thrown in two different situations: (1) token identifier not found; (2) caller != owner.

gRoussac replied:
Let's introduce a new error, please check https://github.com/casper-ecosystem/cep-78-enhanced-nft/pull/181/commits/6fa19fef276a4b496c994f48f2031a421697d437 thanks

gRoussac replied:
I think this time I changed all occurrences from error 28 to 147.

Please check that commit 
https://github.com/casper-ecosystem/cep-78-enhanced-nft/pull/181/commits/7314d36cb43369fbe178eaa3bca1a84c70ed04b8

I had to change a test `should_return_expected_error_when_burning_non_existing_token` that was expecting error 28, now 147.

davidatwhiletrue commented on contract/src/main.rs:
```diff
@@ -807,54 +813,61 @@ pub extern "C" fn approve() {
-    .unwrap_or_revert();
+    let spender = match utils::get_optional_named_arg_with_user_errors::<Key>(
+        ARG_OPERATOR, // Deprecated in favor of ARG_SPENDER
+        NFTCoreError::MissingContractWhiteList,
```
why this error code? should it be `InvalidApprovedAccountHash` ?

gRoussac replied:
Thanks please check that [commit](https://github.com/casper-ecosystem/cep-78-enhanced-nft/pull/181/commits/c632e91fd84f7e4927cd01145d2027b894d25c2d), MissingApprovedAccountHash gets deprecated because ARG_OPERATOR becomes optional. Adding two new errors for explicit spender errors and fixing that copy-paste typo.

ihor commented on README.md:
```diff
@@ -663,14 +671,14 @@ If it is set to `Hash`, you will need to reference the `HASH_BY_INDEX` dictionar
-| Event name      | Included values and type                                   |
-|-----------------|----------------------------------------------------------------------|
+| Event name      | Included values and type                                             |
+| --------------- | -------------------------------------------------------------------- |
```
`token_ids` should be removed from the `ApprovalForAll` event property list

ihor replied:
Separately, the `token_id` should be of type `String`. `Any` type isn't supported by the CES because of decoding issues.

gRoussac replied:
@ihor 
> Is that correct that at the moment there will be no possibility to revoke approval for all and that it will be added later?

There is a possibility to revoke an operator but this was not correctly tested so I added a test below the approve_for_all test to test if revocation as operator occurred correctly.

https://github.com/casper-ecosystem/cep-78-enhanced-nft/pull/181/commits/4cbda35b24509c2e0507c1f9a5271c3c641fa530

gRoussac replied:
https://github.com/casper-ecosystem/cep-78-enhanced-nft/pull/181/commits/092e49d9edb9bd2c645a248c4382e2971e587a8d

davidatwhiletrue commented on contract/src/main.rs:
```diff
@@ -807,54 +813,61 @@ pub extern "C" fn approve() {
+        };
+
+    if !is_owner && !is_operator {
         runtime::revert(NFTCoreError::InvalidAccountHash);
```
should this error code be `InvalidTokenOwner` ?

gRoussac replied:
Thanks I think this is current error code sent for this case and what I tested in corresponding tests.

https://github.com/casper-ecosystem/cep-78-enhanced-nft/blob/feat-cesp/contract/src/main.rs#L827

davidatwhiletrue commented on contract/src/main.rs:
```diff
@@ -866,46 +879,127 @@ pub extern "C" fn approve() {
+        &token_identifier_dictionary_key,
+    ) {
+        Some(owner) => owner,
+        None => runtime::revert(NFTCoreError::InvalidAccountHash),
```
Should this error be `InvalidTokenIdentifier`?

gRoussac replied:
Thanks please see above comment on first occurrence

gRoussac replied:
https://github.com/casper-ecosystem/cep-78-enhanced-nft/pull/181/commits/6fa19fef276a4b496c994f48f2031a421697d437

ihor commented on README.md:
```diff
@@ -680,15 +688,15 @@ For this CEP-78 reference implementation in particular, the events schema is the
-* `NFTIdentifierMode::Ordinal`: the `token id` is encoded as a byte `0x00` followed by a `u64` number.
-* `NFTIdentifierMode::Hash`: the `token_id` is encoded as a byte `0x01` followed by a `String`.
+- `NFTIdentifierMode::Ordinal`: the `token id` is encoded as a byte `0x00` followed by a `u64` number.
+- `NFTIdentifierMode::Hash`: the `token_id` is encoded as a byte `0x01` followed by a `String`.
```
Two paragraphs above should be removed because the `token_id` type was changed to `String`, which can be parsed by a generic parser

gRoussac replied:
Thanks please check that commit, let me know if I missed something

https://github.com/casper-ecosystem/cep-78-enhanced-nft/pull/181/commits/ed86eab5d148af6c9cabbbcb48bd1dac01d21521

gRoussac replied:
https://github.com/casper-ecosystem/cep-78-enhanced-nft/pull/181/commits/092e49d9edb9bd2c645a248c4382e2971e587a8d

davidatwhiletrue commented on contract/src/main.rs:
```diff
@@ -1013,27 +1132,43 @@ pub extern "C" fn transfer() {
+
+    // Revert if caller is not owner nor approved nor an operator.
+    if !is_owner && !is_approved && !is_operator
+    // && !is_operator
```
remove this commented line?

gRoussac replied:
done thanks

ihor commented on contract/src/events/events_ces.rs:
```diff
@@ -43,58 +40,65 @@ impl Burn {
+pub struct ApprovalForAll {
     owner: Key,
     operator: Option<Key>,
+}
```
Since, the `ApprovalForAll` event doesn't have its `RevokeForAll` counterpart then a boolean flag should be added to indicate the action. I would call it `is_approved`.

Or should it be treated as always `true` and the `RevokeForAll` will be added later?

gRoussac replied:
Hi, Please check that commit https://github.com/casper-ecosystem/cep-78-enhanced-nft/pull/181/commits/59ea32601f19c61e4a27a2aafcc0c48ef221034e, please let me know if I missed anything on this new event thanks

gRoussac replied:
https://github.com/casper-ecosystem/cep-78-enhanced-nft/pull/181/commits/135a3fbb64bbccb37d88867410fe0cf455faffa7

gRoussac replied:
Right now I have `RevokedForAll` which is not alike `ApprovalRevoked` nomenclature. Shall I rename to `RevokedForAll` to `ApprovalForAllRevoked` ?

ihor replied:
Let's have `RevokedForAll`, it provides some symmetry with `ApprovalForAll`

davidatwhiletrue commented on README.md:
```diff
@@ -663,14 +671,14 @@ If it is set to `Hash`, you will need to reference the `HASH_BY_INDEX` dictionar
 
-| Event name      | Included values and type                                   |
-|-----------------|----------------------------------------------------------------------|
+| Event name      | Included values and type                                             |
```
Please add here `ApprovalRevoked` event.

gRoussac replied:
Thanks please kindly check that commit https://github.com/casper-ecosystem/cep-78-enhanced-nft/pull/181/commits/ed86eab5d148af6c9cabbbcb48bd1dac01d21521

gRoussac replied:
https://github.com/casper-ecosystem/cep-78-enhanced-nft/pull/181/commits/092e49d9edb9bd2c645a248c4382e2971e587a8d

ihor commented on contract/src/main.rs:
```diff
@@ -914,54 +1008,47 @@ pub extern "C" fn set_approval_for_all() {
+                OwnerReverseLookupMode::NoLookUp => utils::get_owned_token_ids_by_token_number(),
+                OwnerReverseLookupMode::Complete => utils::get_owned_token_ids_by_page(),
+            };
+            for callers_owned_token in caller_owned_tokens {
```
I think here we should also have the `ApprovalForAll` event. The CEP-47 event mode refers to CEP-47 standard event implementation, and to the standard itself.

gRoussac replied:
Hi, I committed this change however I am checking with the rest of the team is this is really what was intended. 

https://github.com/casper-ecosystem/cep-78-enhanced-nft/pull/181/commits/2eee528a7e30d8b44e9451f9abd722ea9b75dfc3

ihor reviewed (changes requested):
@gRoussac please check my comments and questions.

Is that correct that at the moment there will be no possibility to revoke approval for all and that it will be added later?

davidatwhiletrue reviewed (changes requested):
@gRoussac, please see my comments .

gRoussac committed:
- Bad comment
- Missing test on revoke for all

gRoussac committed:
serde version in tests

gRoussac committed:
serde_json in tests like contracts default-features = false

gRoussac committed:
casper-contract in tests default-features = false

gRoussac committed:
Bump some versions

gRoussac committed:
test CI/CD

gRoussac committed:
revert test

gRoussac committed:
Adding new errors + typo + MissingApprovedAccountHash error deprecated

gRoussac committed:
Update doc ApprovalRevoked + token_id as string + removal of token id encoding scheme

gRoussac committed:
token_ids out of ApprovalForAll event

gRoussac committed:
rename encode_dictionary_item_key

gRoussac committed:
stash

gRoussac committed:
Merge branch 'feat-cesp-cherry' into approve_revoke

# Conflicts:
#	contract/Cargo.toml
#	tests/Cargo.toml

gRoussac committed:
Add key_and_value_to_str and fix tests

gRoussac committed:
Introduce RevokeForAll

gRoussac committed:
Add MissingOwnerTokenIdentifierKey

gRoussac committed:
RevokedForAll doc

gRoussac committed:
Add test should_approve_all_with_flat_gas_cost

gRoussac committed:
Merge branch 'feat-cesp' of github.com:casper-ecosystem/cep-78-enhanced-nft into approve_revoke

# Conflicts:
#	contract/Cargo.toml

ACStone-MTS commented on tests/src/burn.rs:
```diff
@@ -544,6 +517,7 @@ fn should_not_burn_in_non_burn_mode() {
     support::assert_expected_error(error, 106, "InvalidBurnMode(106) must have been raised");
 }
 
+// This test is not really relevant anymore as approve_for_all does not check burnt tokens anymore
```
```suggestion
// This test is no longer relevant as approve_for_all does not check burnt tokens anymore
```

Code comment, so probably doesn't need proofreading and you can safely ignore this review comment - just saw it while I was looking through.

ACStone-MTS reviewed (approved):
Documentation portion LGTM

gRoussac committed:
Update tests/src/burn.rs

Co-authored-by: Adam Stone <97986246+ACStoneCL@users.noreply.github.com>

ihor commented on contract/src/main.rs:
```diff
@@ -914,54 +1011,53 @@ pub extern "C" fn set_approval_for_all() {
+    utils::upsert_dictionary_value_from_key(
+        OPERATORS,
+        &owner_operator_item_key,
+        if approve_all { Some(operator) } else { None },
```
Do we need `Some(operator)` here since it's already defined by the dictionary key, we could use store the boolean `approve_all` value here.

gRoussac replied:
Hi, this dictionary stores the key (owner/operator) and Option<Key> as value indeed. I guess it could store a boolean but then we can not query the dictionary from the value (not sure it this makes sense anyway) and I thought this would make more sense to be able to have the direct operator as value here. Let me try with a boolean though as I guess reverse lookup from dictionary value is anyway not possible.

gRoussac replied:
Please check that commit 

https://github.com/casper-ecosystem/cep-78-enhanced-nft/pull/181/commits/7314d36cb43369fbe178eaa3bca1a84c70ed04b8

I am a bit circumspect of the difference between APPROVED and OPERATORS. now, one deal with Option::<Key>, the second with booleans for values which is not super consistent but it works as well. It's probably lighter on storage to save a boolean than an account though.

ihor reviewed (changes requested):
@gRoussac please check my comments

gRoussac committed:
- OPERATORS to bool values
- Missing MissingOwnerTokenIdentifierKey instead of InvalidTokenIdentifier

gRoussac committed:
Merge branch 'approve_revoke' of github.com:gRoussac/cep-78-enhanced-nft into approve_revoke

gRoussac committed:
Remove todo check

gRoussac committed:
Add ApprovalForAll / RevokedForAll to cep47 events

davidatwhiletrue reviewed (approved):
LGTM. Thanks @gRoussac

gRoussac committed:
Merge branch 'feat-cesp' into approve_revoke

# Conflicts:
#	README.md
#	client/get_approved_session/src/main.rs
#	contract/src/error.rs
#	contract/src/main.rs
#	tests/src/burn.rs
#	tests/src/events.rs
#	tests/src/transfer.rs
#	tests/src/upgrade.rs
#	tests/src/utility/constants.rs
#	tests/src/utility/installer_request_builder.rs

gRoussac committed:
namings / cleanup

gRoussac committed:
Update toolchain

gRoussac committed:
Rename function

gRoussac committed:
Deploy fix

ihor reviewed (changes requested):
@gRoussac can we make CEP-47 style event names prefixed with `cep78_` like it was done in the contracts that worked CPE-78 and added CEP-47 style events themselves?

Like in this [example](https://cspr.live/deploy/926c008ba71ec9e45201dd3f7df8b51bfb0d23cce3652acf11c92c959c864cd3): 
<img width="975" alt="Screenshot 2023-03-14 at 20 12 17" src="https://user-images.githubusercontent.com/490943/225112952-3711aaa8-4dc1-46a4-84fe-82599693fa32.png">
